pulsar-broker/src/main/java/org/apache/pulsar/broker/service/GracefulExecutorServicesShutdownHandler.java

```diff
@@ -0,0 +1,116 @@
+ */
+package org.apache.pulsar.broker.service;
+
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.common.util.FutureUtil;
+
+/**
+ * Shuts down one or many {@link ExecutorService}s in a graceful way.
+ *
+ * The executors will be terminated forcefully after the timeout or when the future is cancelled.
+ *
+ * Designed to be used via the API in {@link GracefulExecutorServicesShutdown}
+ */
+@Slf4j
+class GracefulExecutorServicesShutdownHandler {
+    private final ScheduledExecutorService shutdownScheduler = Executors.newSingleThreadScheduledExecutor(
+            new DefaultThreadFactory(getClass().getSimpleName()));
+    private final List<ExecutorService> executors;
+    private final CompletableFuture<Void> future;
+    private final long timeoutMs;
+    private final long shutdownStatusPollingInterval;
+
+    GracefulExecutorServicesShutdownHandler(long timeoutMs, List<ExecutorService> executorServices) {
+        this.timeoutMs = Math.max(timeoutMs, 1L);
+        this.shutdownStatusPollingInterval = Math.min(Math.max(timeoutMs / 100, 10), timeoutMs);
+        executors = executorServices.stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        future = new CompletableFuture<>();
+    }
+
+    CompletableFuture<Void> startShutdownHandler() {
+        log.info("Shutting down {} executors.", executors.size());
+        executors.forEach(ExecutorService::shutdown);
+        FutureUtil.whenCancelledOrTimedOut(future, () -> {
+            terminate();
+        });
+        checkIfExecutorsHaveBeenShutdown();
+        if (!shutdownScheduler.isShutdown()) {
+            try {
+                shutdownScheduler.schedule(this::terminate, timeoutMs, TimeUnit.MILLISECONDS);
+            } catch (RejectedExecutionException e) {
+                // ignore
+            }
+        }
+        return future;
+    }
+
+    private void terminate() {
+        for (ExecutorService executor : executors) {
+            if (!executor.isTerminated()) {
+                log.info("Shutting down forcefully executor {}", executor);
+                for (Runnable runnable : executor.shutdownNow()) {
+                    log.info("Execution in progress for runnable instance of {}: {}", runnable.getClass(),
+                            runnable);
+                }
+            }
+        }
+        shutdown();
+    }
+
+    private void shutdown() {
+        if (!shutdownScheduler.isShutdown()) {
+            log.info("Shutting down scheduler.");
+            shutdownScheduler.shutdown();
```

Comment:
Also for the scheduler thread, shouldn't we await for the thread termination?

Reply:
I'll rewrite the logic to use a `java.lang.Thread` instead of `ScheduledExecutorService`.

Reply:
This is now handled in the rewritten solution.